doc/org-roam.org

```diff
@@ -1380,6 +1201,141 @@ NOTE: Since ripgrep cannot read encrypted files, this function cannot find
+(e.g. how many tasks do I have that are due by next week?). Interfaces such as
+Org-agenda slow to a crawl when the number of files becomes unwieldy, which can
+quickly become the case.
+
+At its core, Org-roam provides a database abstraction layer, providing a dual
+representation of what's already available in plain-text. This allows us
+(humans) to continue working with plain-text, while programs can utilize the
+database layer to perform complex queries. These capabilities include, but are
+not limited to:
+
+- link graph traversal and visualization
+- Instantaneous SQL-like queries on headlines
+  - What are my TODOs, scheduled for X, or due by Y?
+
+All of these functionality is powered by this database abstraction layer. Hence,
+at its core Org-roam's primary goal is to provide a resilient dual
+representation that is cheap to maintain, easy to understand, and is as
+up-to-date as it possibly can. Org-roam also then exposes an API to this
+database abstraction layer for users who would like to perform programmatic
+queries on their Org files.
+
+** Building Extensions and Advanced Customization of Org-roam
+
+Because Org-roam's core functionality is small, it is possible and sometimes
+desirable to build extensions on top of it. These extensions may one or more of
+the following functionalities:
+
+- Access to Org-roam's database
+- Usage/modification of Org-roam's interactive commands
+
+Org-roam provides no guarantees that extensions will continue to function as
+Org-roam evolves, but by following these simple rules, extensions can be made
+robust to local changes in Org-roam.
+
+1. Extensions should not modify the database schema. Any extension that requires
+   the caching of additional data should make a request upstream to Org-roam.
+2. Extensions requiring access to the database should explicitly state support
+   for the database version (~org-roam-db--version~), and only conditionally
+   load when support is available.
+
+*** Accessing the Database
+
+Access to the database is provided singularly by ~org-roam-db-query~, for
+example:
+
+#+begin_src emacs-lisp
+(org-roam-db-query [:select * :from nodes])
+#+end_src
+
+One can refer to the database schema by looking up
+~org-roam-db--table-schemata~. There are multiple helper functions within
+Org-roam that call ~org-roam-db-query~, these are subject to change. To ensure
+that extensions/customizations are robust to change, extensions should only use
+~org-roam-db-query~, and perhaps replicate the SQL query if necessary.
+
+*** Accessing and Modifying Nodes
+
+The node interface is cleanly defined using ~cl-defstruct~. The primary
+method to access nodes is ~org-roam-node-at-point~ and ~org-roam-node-read~:
+
+- Function: org-roam-node-at-point &optional assert
+
+  Return the node at point. If ASSERT, throw an error if there is no node at
+  point.
+
+- Function: org-roam-node-read &optional initial-input filter-fn sort-fn
```

Comment:
BTW, I always was curious, and I think it's time to ask :) . Why `filter-fn` doesn't behave like `filter-fn` in `seq-filter` and instead of taking a `node`, it takes a list of nodes? I understand that this is so much more powerful, but I wonder how real is use-case where someone does some dirty stuff this the list of nodes (e.g. appends new elements).

Reply:
> Why filter-fn doesn't behave like filter-fn in seq-filter and instead of taking a node, it takes a list of nodes?

Probably an oversight of mine while trying to emulate v1's signature, will fix